Northstar.CustomServers/mod/scripts/vscripts/mp/_base_gametype_mp.gnut

```diff
@@ -235,6 +243,43 @@ void function CodeCallback_OnPlayerRespawned( entity player )
 	ClearLastAttacker( player ) // so dying to anything doesn't credit the same attacker after respawning
 }
 
+void function GracePeriodThink( entity player )
+{
+	thread GracePeriodThink_Threaded( player )
+}
+
+void function GracePeriodThink_Threaded( entity player )
+{
+	player.EndSignal( "OnDestroy" )
+	player.EndSignal( "OnPrimaryAttack" )
+	player.EndSignal( "OnDeath" )
+
+	OnThreadEnd(
+		function(): ( player )
+		{
+			if( IsValid( player ) )
+			{
+				if( IsAlive( player ) && GetGameState() == eGameState.Playing ) // if player died we don't reset this( mostly because round end )
+				{
+					player.s.inGracePeriod = false
+					player.Signal( "GracePeriodDone" )
+				}
+			}
+		}
+	)
+
+	float countingTime
+	while( countingTime <= CLASS_CHANGE_GRACE_PERIOD ) // to get exact time
+	{
+		wait 0.2
+		// in dropship
+		if( player.GetParent() != null && ( HasCinematicFlag( player, CE_FLAG_INTRO ) || HasCinematicFlag( player, CE_FLAG_CLASSIC_MP_SPAWNING ) || HasCinematicFlag( player, CE_FLAG_WAVE_SPAWNING ) ) )
+			continue
```

Comment:
This function should use signals to wait for until the player has left the dropship instead of doing this.